### DISCUSSION TITLE
Skin.ini support for beatmap skinning and justice for //Background Colour Transformations

### DISCUSSION THREAD

**Discussion by iBloxman** [ppy/osu#24323] [Ideas]
I've ranked 15+ beatmap skins over the past few years, but one thing I found a bit annoying was the lack of skin.ini support for them. I realize you can set the preferred skin to a custom one, but obviously that requires the user to need to download the skin seperately and gives that annoying popup message if they don't have it which I'd rather avoid.

Just being able to do simple stuff such as SliderBallFlip: 1 so the character in the sliderball doesn't moonwalk backwards every time would be a very welcome change. See https://osu.ppy.sh/beatmapsets/1670776#osu/3423826 or the images below for reference:

![68747470733a2f2f6f73752e7070792e73682f77696b692f696d616765732f536b696e6e696e672f736b696e2e696e692f696d672f536c6964657262616c6c5f666c69702d302e676966](https://github.com/ppy/osu/assets/140199359/e42334cf-e7fd-4a77-ba68-581048e6cc18)
![68747470733a2f2f6f73752e7070792e73682f77696b692f696d616765732f536b696e6e696e672f736b696e2e696e692f696d672f536c6964657262616c6c5f666c69702d312e676966](https://github.com/ppy/osu/assets/140199359/19186901-28e6-496f-bd1d-81ea329c5af6)

Without beatmap skin.ini support, even simple things like having SliderTrackOverride the same color as your combo colors requires you to force preferred skin to default which causes some confusion for aspiring beatmap skinners. Another example, AnimationFramerate would be great to have since I wouldn't need to include include 40 instances of 5 frames of animation to get the exact speed I want the sliderball animation to be, which also bloats the filesize for the map unnecessarily.

There's plenty of other uses I could think of this being helpful for map skinners such as being able to change AllowSliderBallTint, or change the skin version so elements like sliderpoint10.png, show up, etc. I realize beatmap skinning isn't done very much anymore, but it would definitely be a welcome change for the small amount of people who still do it.

I also wanted to breifly mention the //Background Colour Transformations command which seems to currently be broken on stable, and completely missing from Lazer:

 ![SKSS8fY](https://github.com/ppy/osu/assets/140199359/7e899f88-85c8-42ea-915a-9c3f02f79d5c)
 
This issue was breifly mentioned at https://github.com/ppy/osu-stable-issues/issues/125, but as of now it cannot be changed since it returns to its default value after reopening the map. Inputting the command manually into the .osu does not seem to change anything, and the only way for mappers to get this changed is to ask the NAT to manually adjust it.

There are several iconic maps which use these title cards in interesting ways such as timing them on drops in the song or just coloring them to match the background which can give a map a really cool old style feeling. It would just be a shame to ignore these smaller parts of mapping that can make a map really interesting and stand out which can increase the enjoyment for players.

**Comment by peppy** [comment]
Out of curiosity, have you tested beatmap-level skin configuration in lazer with one of the already supported values?

https://github.com/ppy/osu/blob/bf26dbffc2e48458f48d4e8433d2b96f49eacf82/osu.Game/Skinning/SkinConfiguration.cs#L32-L40

I believe these should already work (even when they didn't on stable), and the real issue here is that `SliderBallFlip` is just not implemented at all, but would be great if you could confirm.

**Reply by iBloxman** [reply]
SliderBallFlip seems to be defaulted to 1 on lazer, regardless of what the preferred skins value is set as when importing/exporting the map from stable>lazer. From my testing, it also seems to ignore the "preferred skin" skin.ini values such as SliderBallTint & AnimationFramerate even after the skin in question was imported.

Are beatmap-level skin configuration changes even possible on lazer yet? I'm not sure where to even change these values since the editor doesn't seem to have any of these options, and the skin layout editor doesn't work while editing a map.

These kind of skin configuration options aren't adjustable to my knowledge on stable since we could only change the SliderBorder/SliderTrackOverride under the [Colours] section of the .osu file. The only alternative being making a custom skin & making it the preferred skin which introduces that "beatmap is requesting a skin popup" if players don't have it.

**Reply by peppy** [reply]
> Are beatmap-level skin configuration changes even possible on lazer yet?  

That's what i'm asking you to test.

> I'm not sure where to even change these values since the editor doesn't seem to have any of these options, and the skin layout editor doesn't work while editing a map.

In the .osu file, using the same definitions as `skin.ini`.

**Reply by iBloxman** [reply]
After saving and exporting the .osz from stable to lazer, it doesn't appear to change SliderballTint, SliderBallFlip, or AnimationFramerate regardless of where I put them in the .osu file. I've tried putting the commands individually under the [Colours] and [General] tabs to see if that makes any difference, but neither seem to work.

Commands under the [General] tab seem to disappear or revert back to the live version of the map even after being exported. Commands under the [Colours] tab (other than SliderBorder/TrackOverride) seem to instantly make the diff unreadable.

![image](https://github.com/ppy/osu/assets/140199359/e3e7cc47-3a67-4cb7-8b72-d6696d54bd82)

Even if I edit as a new map and export with a freshly placed slider object (or just switch to another difficulty and export the .osz), it does not change either skin setting on lazer regardless.

> That's what i'm asking you to test.

I had thought you meant there was a way to change the skin configuration settings _within_ lazer, which I was confused about since that is what I was suggesting being implemented.

**Reply by peppy** [reply]
> Commands under the [General] tab seem to disappear or revert back to the live version of the map even after being exported

You'd need to make the changes without exporting. Never mind, I'll test myself when we get around to looking into this.

**Comment by fluxie31** [comment]
> > Commands under the [General] tab seem to disappear or revert back to the live version of the map even after being exported
> 
> You'd need to make the changes without exporting. Never mind, I'll test myself when we get around to looking into this.

Hey, do you want me to test it, if it hasn't already been tested? I don't see any updates past July so I just wanted to ask
